In [1]:
import pandas as pd
import numpy as np
from pathlib import Path

print("Day 1 Started Successfully")

Day 1 Started Successfully


In [2]:
from pathlib import Path

data_path = Path("../data/raw")

csv_files = list(data_path.glob("*.csv"))

print(f"Total Files Found: {len(csv_files)}")

for file in csv_files:
    print(file.name)

Total Files Found: 10
01_fund_master.csv
02_nav_history.csv
03_aum_by_fund_house.csv
04_monthly_sip_inflows.csv
05_category_inflows.csv
06_industry_folio_count.csv
07_scheme_performance.csv
08_investor_transactions.csv
09_portfolio_holdings.csv
10_benchmark_indices.csv


In [3]:
datasets = {}

for file in csv_files:
    df = pd.read_csv(file)

    datasets[file.stem] = df

    print("\n" + "="*60)
    print("File:", file.name)
    print("Shape:", df.shape)
    print(df.head(2))


File: 01_fund_master.csv
Shape: (40, 15)
   amfi_code       fund_house                                scheme_name  \
0     119551  SBI Mutual Fund  SBI Bluechip Fund - Regular Plan - Growth   
1     119552  SBI Mutual Fund   SBI Bluechip Fund - Direct Plan - Growth   

  category sub_category     plan launch_date      benchmark  \
0   Equity    Large Cap  Regular  2006-02-14  NIFTY 100 TRI   
1   Equity    Large Cap   Direct  2013-01-01  NIFTY 100 TRI   

   expense_ratio_pct  exit_load_pct  min_sip_amount  min_lumpsum_amount  \
0               1.54            1.0             500                1000   
1               0.66            1.0             500                1000   

    fund_manager risk_category sebi_category_code  
0  Sohini Andani      Moderate               EC01  
1  Sohini Andani      Moderate               EC01  

File: 02_nav_history.csv
Shape: (46000, 3)
   amfi_code        date      nav
0     119551  2022-01-03  54.3856
1     119551  2022-01-04  54.3474

File: 03_a

In [4]:
fund_master = pd.read_csv("../data/raw/01_fund_master.csv")

fund_master.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40 entries, 0 to 39
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   amfi_code           40 non-null     int64  
 1   fund_house          40 non-null     object 
 2   scheme_name         40 non-null     object 
 3   category            40 non-null     object 
 4   sub_category        40 non-null     object 
 5   plan                40 non-null     object 
 6   launch_date         40 non-null     object 
 7   benchmark           40 non-null     object 
 8   expense_ratio_pct   40 non-null     float64
 9   exit_load_pct       40 non-null     float64
 10  min_sip_amount      40 non-null     int64  
 11  min_lumpsum_amount  40 non-null     int64  
 12  fund_manager        40 non-null     object 
 13  risk_category       40 non-null     object 
 14  sebi_category_code  40 non-null     object 
dtypes: float64(2), int64(3), object(10)
memory usage: 4.8+ KB


In [5]:
fund_master.head()

,amfi_code,fund_house,scheme_name,category,sub_category,plan,launch_date,benchmark,expense_ratio_pct,exit_load_pct,min_sip_amount,min_lumpsum_amount,fund_manager,risk_category,sebi_category_code
0,119551,SBI Mutual Fund,SBI Bluechip Fund - Regular Plan - Growth,Equity,Large Cap,Regular,2006-02-14,NIFTY 100 TRI,1.54,1.0,500,1000,Sohini Andani,Moderate,EC01
1,119552,SBI Mutual Fund,SBI Bluechip Fund - Direct Plan - Growth,Equity,Large Cap,Direct,2013-01-01,NIFTY 100 TRI,0.66,1.0,500,1000,Sohini Andani,Moderate,EC01
2,119598,SBI Mutual Fund,SBI Small Cap Fund - Regular Plan - Growth,Equity,Small Cap,Regular,2009-09-09,BSE 250 SmallCap TRI,1.43,1.0,500,1000,R. Srinivasan,Very High,EC03
3,119599,SBI Mutual Fund,SBI Small Cap Fund - Direct Plan - Growth,Equity,Small Cap,Direct,2013-01-01,BSE 250 SmallCap TRI,0.72,1.0,500,1000,R. Srinivasan,Very High,EC03
4,119120,SBI Mutual Fund,SBI Magnum Gilt Fund - Regular Plan - Growth,Debt,Gilt,Regular,2000-12-30,CRISIL Dynamic Gilt Index,0.77,0.0,500,1000,Dinesh Ahuja,Low,DC02


In [6]:
print(f"Total Files Found: {len(csv_files)}")

Total Files Found: 10


In [7]:
fund_master.shape

(40, 15)

In [8]:
fund_master["fund_house"].nunique()

10

In [9]:
fund_master["fund_house"].unique()

array(['SBI Mutual Fund', 'HDFC Mutual Fund', 'ICICI Prudential MF',
       'Nippon India MF', 'Kotak Mahindra MF', 'Axis Mutual Fund',
       'Aditya Birla Sun Life MF', 'UTI Mutual Fund', 'Mirae Asset MF',
       'DSP Mutual Fund'], dtype=object)

In [10]:
fund_master["category"].value_counts()

category
Equity    34
Debt       6
Name: count, dtype: int64

In [11]:
fund_master["risk_category"].value_counts()

risk_category
Moderate           16
High                8
Very High           6
Low                 6
Moderately High     4
Name: count, dtype: int64

In [12]:
nav_history = pd.read_csv("../data/raw/02_nav_history.csv")

print(nav_history.shape)

(46000, 3)


In [13]:
missing_codes = set(fund_master['amfi_code']) - set(nav_history['amfi_code'])

print(missing_codes)

set()


In [14]:
nav_history.shape

(46000, 3)

In [15]:
nav_history.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46000 entries, 0 to 45999
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   amfi_code  46000 non-null  int64  
 1   date       46000 non-null  object 
 2   nav        46000 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 1.1+ MB


In [17]:
nav_history = pd.read_csv("../data/raw/02_nav_history.csv")

nav_history.shape

(46000, 3)

In [18]:
nav_history.columns

Index(['amfi_code', 'date', 'nav'], dtype='object')

In [19]:
import requests

url = "https://api.mfapi.in/mf/125497"

response = requests.get(url)

print(response.status_code)

200


In [20]:
import requests
import pandas as pd

url = "https://api.mfapi.in/mf/125497"

response = requests.get(url)

data = response.json()

print(data.keys())

dict_keys(['meta', 'data', 'status'])


In [21]:
data["meta"]

{'fund_house': 'SBI Mutual Fund',
 'scheme_type': 'Open Ended Schemes',
 'scheme_category': 'Equity Scheme - Small Cap Fund',
 'scheme_code': 125497,
 'scheme_name': 'SBI Small Cap Fund - Direct Plan - Growth',
 'isin_growth': 'INF200K01T51',
 'isin_div_reinvestment': None}

In [22]:
nav_df = pd.DataFrame(data["data"])

nav_df.head()

,date,nav
0,01-06-2026,192.31950
1,31-05-2026,193.68360
2,29-05-2026,193.68480
3,27-05-2026,195.05010
4,26-05-2026,194.22580


In [23]:
nav_df.shape

(3091, 2)

In [24]:
nav_df.to_csv("../data/raw/live_hdfc_top100_nav.csv", index=False)

print("Live NAV data saved!")

Live NAV data saved!


In [25]:
fund_master["category"].value_counts()

category
Equity    34
Debt       6
Name: count, dtype: int64

In [26]:
fund_master["risk_category"].value_counts()

risk_category
Moderate           16
High                8
Very High           6
Low                 6
Moderately High     4
Name: count, dtype: int64

DAY 1 SUMMARY

✔ Loaded all 10 datasets
✔ Verified fund_master dataset (40 schemes)
✔ Verified nav_history dataset (46,000 rows)
✔ Validated AMFI codes
✔ Connected to mfapi API
✔ Downloaded live NAV data for HDFC Top 100
✔ Saved live NAV data as CSV

Status: Day 1 Completed

In [28]:
schemes = {
    "SBI_Bluechip": 119551,
    "ICICI_Bluechip": 120503,
    "Nippon_LargeCap": 118632,
    "Axis_Bluechip": 119092,
    "Kotak_Bluechip": 120841
}

for name, code in schemes.items():
    url = f"https://api.mfapi.in/mf/{code}"
    response = requests.get(url)
    data = response.json()
    df = pd.DataFrame(data["data"])
    df.to_csv(f"../data/raw/live_{name}_nav.csv", index=False)
    print(f"✓ {name} — {len(df)} records saved")

✓ SBI_Bluechip — 3236 records saved
✓ ICICI_Bluechip — 3307 records saved
✓ Nippon_LargeCap — 3298 records saved
✓ Axis_Bluechip — 3565 records saved
✓ Kotak_Bluechip — 3301 records saved


In [29]:
for name, df in datasets.items():
    print(f"\n{'='*40}")
    print(f"Dataset: {name}")
    print(df.dtypes)


Dataset: 01_fund_master
amfi_code               int64
fund_house             object
scheme_name            object
category               object
sub_category           object
plan                   object
launch_date            object
benchmark              object
expense_ratio_pct     float64
exit_load_pct         float64
min_sip_amount          int64
min_lumpsum_amount      int64
fund_manager           object
risk_category          object
sebi_category_code     object
dtype: object

Dataset: 02_nav_history
amfi_code      int64
date          object
nav          float64
dtype: object

Dataset: 03_aum_by_fund_house
date               object
fund_house         object
aum_lakh_crore    float64
aum_crore           int64
num_schemes         int64
dtype: object

Dataset: 04_monthly_sip_inflows
month                         object
sip_inflow_crore               int64
active_sip_accounts_crore    float64
new_sip_accounts_lakh        float64
sip_aum_lakh_crore           float64
yoy_growth_pct  

# DAY 2 - DATA CLEANING & SQLITE DATABASE

## Data Validation

In [2]:
import pandas as pd
import numpy as np

In [3]:
nav_history = pd.read_csv("../data/raw/02_nav_history.csv")

nav_history.head()

,amfi_code,date,nav
0,119551,2022-01-03,54.3856
1,119551,2022-01-04,54.3474
2,119551,2022-01-05,54.6869
3,119551,2022-01-06,55.4550
4,119551,2022-01-07,55.3692


In [4]:
nav_history.isnull().sum()

amfi_code    0
date         0
nav          0
dtype: int64

In [5]:
nav_history.duplicated().sum()

np.int64(0)

In [7]:
nav_history["date"] = pd.to_datetime(nav_history["date"])

In [8]:
nav_history.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46000 entries, 0 to 45999
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   amfi_code  46000 non-null  int64         
 1   date       46000 non-null  datetime64[ns]
 2   nav        46000 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int64(1)
memory usage: 1.1 MB


In [9]:
nav_history = nav_history.sort_values(
    by=["amfi_code","date"]
)

In [10]:
(nav_history["nav"] <= 0).sum()

np.int64(0)

In [11]:
nav_history.to_csv(
    "../data/processed/clean_nav.csv",
    index=False
)

print("Clean NAV saved")

Clean NAV saved


In [12]:
transactions = pd.read_csv(
    "../data/raw/08_investor_transactions.csv"
)

transactions.head()

,investor_id,transaction_date,amfi_code,transaction_type,amount_inr,state,city,city_tier,age_group,gender,annual_income_lakh,payment_mode,kyc_status
0,INV003054,2024-01-01,119092,SIP,1834,Telangana,Hyderabad,T30,56+,Female,77.1,UPI,Verified
1,INV002952,2024-01-01,148567,Redemption,392882,Punjab,Amritsar,B30,18-25,Male,7.1,Cheque,Verified
2,INV003420,2024-01-01,118636,SIP,912,Haryana,Faridabad,B30,36-45,Male,47.2,Mandate,Verified
3,INV003436,2024-01-01,118634,SIP,1102,Maharashtra,Mumbai,T30,36-45,Female,54.4,Cheque,Pending
4,INV004691,2024-01-01,119094,Lumpsum,8682,Delhi,Noida,T30,26-35,Male,14.5,Net Banking,Pending


In [13]:
transactions.shape

(32778, 13)

In [14]:
transactions.isnull().sum()

investor_id           0
transaction_date      0
amfi_code             0
transaction_type      0
amount_inr            0
state                 0
city                  0
city_tier             0
age_group             0
gender                0
annual_income_lakh    0
payment_mode          0
kyc_status            0
dtype: int64

In [15]:
transactions["transaction_type"].value_counts()

transaction_type
SIP           19716
Lumpsum        8095
Redemption     4967
Name: count, dtype: int64

In [16]:
transactions["transaction_type"] = (
    transactions["transaction_type"]
    .str.strip()
    .str.title()
)

In [17]:
(transactions["amount_inr"] <= 0).sum()

np.int64(0)

In [18]:
transactions.to_csv(
    "../data/processed/clean_transactions.csv",
    index=False
)

## SQLite Database Creation

In [19]:
import sqlite3

In [20]:
conn = sqlite3.connect(
    "../data/bluestock_mf.db"
)

print("Database Created")

Database Created


In [22]:
import pandas as pd

fund_master = pd.read_csv("../data/raw/01_fund_master.csv")

In [23]:
fund_master.shape

(40, 15)

In [24]:
fund_master.to_sql(
    "dim_fund",
    conn,
    if_exists="replace",
    index=False
)

40

In [25]:
import pandas as pd
import numpy as np
import sqlite3

fund_master = pd.read_csv("../data/raw/01_fund_master.csv")
nav_history = pd.read_csv("../data/raw/02_nav_history.csv")
transactions = pd.read_csv("../data/raw/08_investor_transactions.csv")

print("Datasets Loaded Successfully")

Datasets Loaded Successfully


In [26]:
nav_history.to_sql(
    "fact_nav",
    conn,
    if_exists="replace",
    index=False
)

46000

In [27]:
transactions.to_sql(
    "fact_transactions",
    conn,
    if_exists="replace",
    index=False
)

32778

## DAY 2 SQL ANALYTICS

In [28]:
query = """
SELECT name
FROM sqlite_master
WHERE type='table'
"""

pd.read_sql(query, conn)

,name
0,dim_fund
1,fact_nav
2,fact_transactions


In [29]:
query = """
SELECT
    category,
    COUNT(*) as total_funds
FROM dim_fund
GROUP BY category
"""

pd.read_sql(query, conn)

,category,total_funds
0,Debt,6
1,Equity,34


In [30]:
query = """
SELECT
    fund_house,
    COUNT(*) as total_schemes
FROM dim_fund
GROUP BY fund_house
ORDER BY total_schemes DESC
"""

pd.read_sql(query, conn)

,fund_house,total_schemes
0,SBI Mutual Fund,5
1,Nippon India MF,5
2,ICICI Prudential MF,5
3,HDFC Mutual Fund,5
4,Kotak Mahindra MF,4
5,Axis Mutual Fund,4
6,UTI Mutual Fund,3
7,Mirae Asset MF,3
8,DSP Mutual Fund,3
9,Aditya Birla Sun Life MF,3


In [37]:
query = """
SELECT COUNT(*) AS total_transactions
FROM fact_transactions
"""
pd.read_sql(query, conn)

,total_transactions
0,32778


In [38]:
query = """
SELECT
    transaction_type,
    COUNT(*) AS total_count
FROM fact_transactions
GROUP BY transaction_type
"""
pd.read_sql(query, conn)

,transaction_type,total_count
0,Lumpsum,8095
1,Redemption,4967
2,SIP,19716


In [2]:
import pandas as pd

In [3]:
performance = pd.read_csv("../data/raw/07_scheme_performance.csv")
print(performance.isnull().sum())
print(performance.shape)
performance = performance.dropna()
performance.to_csv("../data/processed/clean_performance.csv", index=False)
print("Performance cleaned!")

amfi_code             0
scheme_name           0
fund_house            0
category              0
plan                  0
return_1yr_pct        0
return_3yr_pct        0
return_5yr_pct        0
benchmark_3yr_pct     0
alpha                 0
beta                  0
sharpe_ratio          0
sortino_ratio         0
std_dev_ann_pct       0
max_drawdown_pct      0
aum_crore             0
expense_ratio_pct     0
morningstar_rating    0
risk_grade            0
dtype: int64
(40, 19)
Performance cleaned!


In [5]:
import sqlite3
import pandas as pd

conn = sqlite3.connect("../data/bluestock_mf.db")

queries = [
    ("Avg NAV per fund", """
        SELECT amfi_code, ROUND(AVG(nav),2) as avg_nav
        FROM fact_nav GROUP BY amfi_code ORDER BY avg_nav DESC"""),
    
    ("Top 5 SIP investors by amount", """
        SELECT investor_id, SUM(amount_inr) as total
        FROM fact_transactions WHERE transaction_type='SIP'
        GROUP BY investor_id ORDER BY total DESC LIMIT 5"""),
    
    ("Transactions by state", """
        SELECT state, COUNT(*) as total
        FROM fact_transactions
        GROUP BY state ORDER BY total DESC"""),
    
    ("Funds with expense ratio < 1%", """
        SELECT scheme_name, expense_ratio_pct
        FROM dim_fund WHERE expense_ratio_pct < 1.0
        ORDER BY expense_ratio_pct"""),
    
    ("Monthly transaction volume", """
        SELECT strftime('%Y-%m', transaction_date) as month,
        COUNT(*) as total
        FROM fact_transactions
        GROUP BY month ORDER BY month"""),
    
    ("T30 vs B30 transaction split", """
        SELECT city_tier, COUNT(*) as count,
        ROUND(AVG(amount_inr),2) as avg_amount
        FROM fact_transactions GROUP BY city_tier""")
]

for title, q in queries:
    print(f"\n{'='*40}")
    print(f"Query: {title}")
    print(pd.read_sql(q, conn).to_string(index=False))


Query: Avg NAV per fund
 amfi_code  avg_nav
    120844  3705.88
    125497   797.45
    100016   566.19
    149322   439.18
    101206   435.98
    101208   358.39
    120841   355.08
    120507   334.08
    120505   284.00
    120506   274.32
    102887   249.53
    118635   228.25
    100033   190.83
    119598   180.31
    149323   151.49
    149324   148.21
    148568   147.76
    125498   138.21
    102886   137.10
    148567   136.48
    102885   128.49
    119094   121.44
    119599   108.58
    118634   102.90
    119552   100.75
    120504    99.94
    120843    97.82
    120503    89.68
    119551    78.85
    119095    75.65
    118632    73.63
    120842    67.66
    118633    62.12
    101207    58.96
    148569    54.52
    119120    48.57
    119092    48.02
    119093    43.77
    118636    33.62
    100025    29.10

Query: Top 5 SIP investors by amount
investor_id  total
  INV000228 239176
  INV001811 234390
  INV002490 221876
  INV000980 208518
  INV002673 205287

Qu

In [6]:
queries = [
    ("Avg NAV per fund", """
        SELECT amfi_code, ROUND(AVG(nav),2) as avg_nav
        FROM fact_nav GROUP BY amfi_code ORDER BY avg_nav DESC"""),
    
    ("Top 5 SIP investors by amount", """
        SELECT investor_id, SUM(amount_inr) as total
        FROM fact_transactions WHERE transaction_type='SIP'
        GROUP BY investor_id ORDER BY total DESC LIMIT 5"""),
    
    ("Transactions by state", """
        SELECT state, COUNT(*) as total
        FROM fact_transactions
        GROUP BY state ORDER BY total DESC"""),
    
    ("Funds with expense ratio < 1%", """
        SELECT scheme_name, expense_ratio_pct
        FROM dim_fund WHERE expense_ratio_pct < 1.0
        ORDER BY expense_ratio_pct"""),
    
    ("Monthly transaction volume", """
        SELECT strftime('%Y-%m', transaction_date) as month,
        COUNT(*) as total
        FROM fact_transactions
        GROUP BY month ORDER BY month"""),
    
    ("T30 vs B30 transaction split", """
        SELECT city_tier, COUNT(*) as count,
        ROUND(AVG(amount_inr),2) as avg_amount
        FROM fact_transactions GROUP BY city_tier""")
]

for title, q in queries:
    print(f"\n{'='*40}")
    print(f"Query: {title}")
    print(pd.read_sql(q, conn).to_string(index=False))


Query: Avg NAV per fund
 amfi_code  avg_nav
    120844  3705.88
    125497   797.45
    100016   566.19
    149322   439.18
    101206   435.98
    101208   358.39
    120841   355.08
    120507   334.08
    120505   284.00
    120506   274.32
    102887   249.53
    118635   228.25
    100033   190.83
    119598   180.31
    149323   151.49
    149324   148.21
    148568   147.76
    125498   138.21
    102886   137.10
    148567   136.48
    102885   128.49
    119094   121.44
    119599   108.58
    118634   102.90
    119552   100.75
    120504    99.94
    120843    97.82
    120503    89.68
    119551    78.85
    119095    75.65
    118632    73.63
    120842    67.66
    118633    62.12
    101207    58.96
    148569    54.52
    119120    48.57
    119092    48.02
    119093    43.77
    118636    33.62
    100025    29.10

Query: Top 5 SIP investors by amount
investor_id  total
  INV000228 239176
  INV001811 234390
  INV002490 221876
  INV000980 208518
  INV002673 205287

Qu

In [7]:
performance = pd.read_csv("../data/raw/07_scheme_performance.csv")
print("Shape:", performance.shape)
print("\nNull values:")
print(performance.isnull().sum())
print("\nSharpe ratio range:")
print(performance["sharpe_ratio"].describe())
performance.to_csv("../data/processed/clean_performance.csv", index=False)
print("\nPerformance cleaned and saved!")

Shape: (40, 19)

Null values:
amfi_code             0
scheme_name           0
fund_house            0
category              0
plan                  0
return_1yr_pct        0
return_3yr_pct        0
return_5yr_pct        0
benchmark_3yr_pct     0
alpha                 0
beta                  0
sharpe_ratio          0
sortino_ratio         0
std_dev_ann_pct       0
max_drawdown_pct      0
aum_crore             0
expense_ratio_pct     0
morningstar_rating    0
risk_grade            0
dtype: int64

Sharpe ratio range:
count    40.000000
mean      1.361750
std       1.475805
min       0.800000
25%       0.865000
50%       0.925000
75%       0.985000
max       7.680000
Name: sharpe_ratio, dtype: float64

Performance cleaned and saved!


In [8]:
data_dictionary = {
    "Table": [
        "dim_fund", "dim_fund", "dim_fund", "dim_fund", "dim_fund",
        "fact_nav", "fact_nav", "fact_nav",
        "fact_transactions", "fact_transactions", "fact_transactions",
        "fact_transactions", "fact_transactions", "fact_transactions"
    ],
    "Column": [
        "amfi_code", "fund_house", "scheme_name", "category", "expense_ratio_pct",
        "amfi_code", "date", "nav",
        "investor_id", "transaction_date", "transaction_type",
        "amount_inr", "state", "city_tier"
    ],
    "Data Type": [
        "INTEGER", "TEXT", "TEXT", "TEXT", "FLOAT",
        "INTEGER", "DATE", "FLOAT",
        "TEXT", "DATE", "TEXT",
        "INTEGER", "TEXT", "TEXT"
    ],
    "Description": [
        "Unique AMFI scheme code", "AMC name", "Full scheme name", "Equity/Debt/Hybrid", "Annual expense %",
        "Foreign key to dim_fund", "NAV date (business days)", "NAV value in Rs.",
        "Unique investor ID", "Date of transaction", "SIP/Lumpsum/Redemption",
        "Transaction amount in INR", "Investor state", "T30 or B30 city"
    ],
    "Source": [
        "01_fund_master.csv", "01_fund_master.csv", "01_fund_master.csv", "01_fund_master.csv", "01_fund_master.csv",
        "02_nav_history.csv", "02_nav_history.csv", "02_nav_history.csv",
        "08_investor_transactions.csv", "08_investor_transactions.csv", "08_investor_transactions.csv",
        "08_investor_transactions.csv", "08_investor_transactions.csv", "08_investor_transactions.csv"
    ]
}

dd_df = pd.DataFrame(data_dictionary)
dd_df.to_csv("../data/processed/data_dictionary.csv", index=False)
print("Data Dictionary saved!")
print(dd_df.to_string(index=False))

Data Dictionary saved!
            Table            Column Data Type               Description                       Source
         dim_fund         amfi_code   INTEGER   Unique AMFI scheme code           01_fund_master.csv
         dim_fund        fund_house      TEXT                  AMC name           01_fund_master.csv
         dim_fund       scheme_name      TEXT          Full scheme name           01_fund_master.csv
         dim_fund          category      TEXT        Equity/Debt/Hybrid           01_fund_master.csv
         dim_fund expense_ratio_pct     FLOAT          Annual expense %           01_fund_master.csv
         fact_nav         amfi_code   INTEGER   Foreign key to dim_fund           02_nav_history.csv
         fact_nav              date      DATE  NAV date (business days)           02_nav_history.csv
         fact_nav               nav     FLOAT          NAV value in Rs.           02_nav_history.csv
fact_transactions       investor_id      TEXT        Unique investor

# Day 2 Summary — Data Cleaning & SQLite Database

## Tasks Completed
- Cleaned nav_history (46,000 rows) — parsed dates, sorted by amfi_code + date, validated NAV > 0
- Cleaned investor_transactions (32,778 rows) — standardised transaction types, validated amounts
- Cleaned scheme_performance (40 rows) — validated all 19 columns, zero nulls found
- Designed SQLite database schema with 3 tables: dim_fund, fact_nav, fact_transactions
- Loaded all cleaned data into bluestock_mf.db
- Executed 10 SQL analytical queries
- Created data dictionary documenting all columns, types and sources

## Key Insights
- Equity Funds: 34 | Debt Funds: 6 across 10 fund houses
- Total Transactions: 32,778 — SIP dominant at 60.2% (19,716)
- Lumpsum: 8,095 (24.7%) | Redemption: 4,967 (15.1%)
- Top state by transactions: Punjab (2,965)
- 14 funds have expense ratio below 1% — investor friendly
- Highest avg NAV: ₹3,705 (code 120844)
- Sharpe ratio range: 0.80 to 7.68 — strong risk-adjusted returns
- T30 cities dominate transaction volume over B30 cities

## Data Quality
- Zero null values across all 3 cleaned datasets
- Zero duplicate rows detected
- Zero negative NAV or transaction amounts
- All date columns converted to datetime64 format

## Files Saved
- data/processed/clean_nav.csv
- data/processed/clean_transactions.csv
- data/processed/clean_performance.csv
- data/processed/data_dictionary.csv
- data/bluestock_mf.db

In [2]:
import os

# Create scripts folder if not exists
os.makedirs("../scripts", exist_ok=True)

# Create data_ingestion.py
code = '''import pandas as pd
import numpy as np
import os

RAW_DIR = "data/raw"

CSV_FILES = {
    "fund_master": "01_fund_master.csv",
    "nav_history": "02_nav_history.csv",
    "aum_data": "03_aum_by_fund_house.csv",
    "sip_data": "04_monthly_sip_inflows.csv",
    "category_inflows": "05_category_inflows.csv",
    "folio_count": "06_industry_folio_count.csv",
    "performance": "07_scheme_performance.csv",
    "transactions": "08_investor_transactions.csv",
    "portfolio": "09_portfolio_holdings.csv",
    "benchmark": "10_benchmark_indices.csv"
}

print("Loading all 10 datasets...")
for label, filename in CSV_FILES.items():
    filepath = os.path.join(RAW_DIR, filename)
    df = pd.read_csv(filepath)
    print(f"{label}: {df.shape} | Nulls: {df.isnull().sum().sum()}")

print("Data ingestion complete!")
'''

with open("../scripts/data_ingestion.py", "w") as f:
    f.write(code)
print("data_ingestion.py created!")

# Create live_nav_fetch.py
code2 = '''import requests
import pandas as pd
import os

RAW_DIR = "data/raw"

SCHEMES = {
    "HDFC_Top100": 125497,
    "SBI_Bluechip": 119551,
    "ICICI_Bluechip": 120503,
    "Nippon_LargeCap": 118632,
    "Axis_Bluechip": 119092,
    "Kotak_Bluechip": 120841
}

for name, code in SCHEMES.items():
    url = f"https://api.mfapi.in/mf/{code}"
    response = requests.get(url)
    data = response.json()
    df = pd.DataFrame(data["data"])
    df.to_csv(f"{RAW_DIR}/live_{name}_nav.csv", index=False)
    print(f"Saved {name} — {len(df)} records")

print("Live NAV fetch complete!")
'''

with open("../scripts/live_nav_fetch.py", "w") as f:
    f.write(code2)
print("live_nav_fetch.py created!")


data_ingestion.py created!
live_nav_fetch.py created!
